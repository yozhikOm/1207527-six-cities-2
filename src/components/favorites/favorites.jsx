import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {Header} from '../header/header.jsx';
import {getCities} from '../../reducer/data/selectors.js';
import {PropertyCard} from '../property-card/property-card.jsx';

class Favorites extends PureComponent {

  componentDidMount() {
    this.props.loadFavorites();
  }

  render() {
    const {favorites, isAuthorizationRequired, userInfo, onSetActiveItem} = this.props;

    if (!favorites) {
      return (
        <div>Loading...</div>
      );
    } else {

      const cities = getCities(favorites);

      return (
        <div className="page page--gray page--main">
          <Header isAuthorizationRequired={isAuthorizationRequired} userInfo={userInfo} />

          <main className="page__main page__main--favorites">
            <div className="page__favorites-container container">
              {favorites.length === 0 ?
                (<section className="favorites favorites--empty">
                  <h1 className="visually-hidden">Favorites (empty)</h1>
                  <div className="favorites__status-wrapper">
                    <b className="favorites__status">Nothing yet saved.</b>
                    <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                  </div>
                </section>)
                :
                (<section className="favorites">
                  <h1 className="favorites__title">Saved listing</h1>
                  <ul className="favorites__list">
                    {cities.map((it, i) => (
                      <li className="favorites__locations-items" key={`fav-${i}`}>
                        <div className="favorites__locations locations locations--current">
                          <div className="locations__item">
                            <a className="locations__item-link" href="#">
                              <span>{it.title}</span>
                            </a>
                          </div>
                        </div>
                        <div className="favorites__places">
                          {favorites.filter((f) => f.city.name === it.title).map((offer) => (
                            <React.Fragment key={offer.id}>
                              <PropertyCard offer={offer} cardMouseEnterHandler={onSetActiveItem}/>
                            </React.Fragment>
                          ))}
                        </div>
                      </li>
                    ))}
                  </ul>
                </section>)
              }

            </div>
          </main>
          <footer className="footer container">
            <a className="footer__logo-link" href="main.html">
              <img className="footer__logo" src="/img/logo.svg" alt="6 cities logo" width="64" height="33" />
            </a>
          </footer>
        </div>
      );
    }
  }
}

Favorites.propTypes = {
  history: PropTypes.object,
  isAuthorizationRequired: PropTypes.bool,
  userInfo: PropTypes.shape({
    id: PropTypes.id,
    name: PropTypes.string,
    avatarUrl: PropTypes.string,
    isPro: PropTypes.bool
  }),
  loadFavorites: PropTypes.func,
  favorites: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    city: PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number,
      }),
    }).isRequired,
    previewImage: PropTypes.string,
    images: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool,
    isPremium: PropTypes.bool,
    rating: PropTypes.number,
    type: PropTypes.string.isRequired,
    bedrooms: PropTypes.number,
    maxAdults: PropTypes.number,
    price: PropTypes.number.isRequired,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      isPro: PropTypes.bool,
      avatarUrl: PropTypes.string,
    }),
    description: PropTypes.string.isRequired,
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number,
    }),
  })),
  onSetActiveItem: PropTypes.func,
};

export {Favorites};
