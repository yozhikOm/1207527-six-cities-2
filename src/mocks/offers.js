export const offers = [
  {
    "id": 1,
    "title": `Cozy house amsterdam`,
    "type": `House`,
    "price": 650,
    "photos": [
      {
        "src": `img/apartment-01.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3909553943508, 4.85309666406198],
  },
  {
    "id": 2,
    "title": `Cosy room in spacious apartment`,
    "type": `Room`,
    "price": 45,
    "photos": [
      {
        "src": `img/apartment-02.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.369553943508, 4.85309666406198],
  },
  {
    "id": 3,
    "title": `Elegant private room in stylish house`,
    "type": `Room`,
    "price": 128,
    "photos": [
      {
        "src": `img/room.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3909553943508, 4.929309666406198],
  },
  {
    "id": 4,
    "title": `Huge Tinyhouse in the green, between sea and city`,
    "type": `House`,
    "price": 1520,
    "photos": [
      {
        "src": `img/apartment-03.jpg`
      },
      {
        "src": `img/amsterdam.jpg`
      },
    ],
    "coordinates": [52.3809553943508, 4.939309666406198],
  }
];
