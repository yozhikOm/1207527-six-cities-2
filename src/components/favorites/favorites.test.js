import React from 'react';
import {Favorites} from './favorites.jsx';
import ShallowRenderer from 'react-test-renderer/shallow';
import {offers} from '../../mocks/offers.js';

it(`Favorites компонент рендерится корректно`, () => {
  const renderer = new ShallowRenderer();

  const mockUserInfo = {
    id: 2,
    name: `fakeUser`,
    avatarUrl: ``,
    isPro: true
  };

  renderer.render(<Favorites
    favorites={offers}
    isAuthorizationRequired={false}
    userInfo={mockUserInfo}
    onSetActiveItem={jest.fn()}
  />);

  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
