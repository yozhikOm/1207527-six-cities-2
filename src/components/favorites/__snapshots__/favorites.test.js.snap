// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites компонент рендерится корректно 1`] = `
<div
  className="page page--gray page--main"
>
  <Header
    isAuthorizationRequired={false}
    userInfo={
      Object {
        "avatarUrl": "",
        "id": 2,
        "isPro": true,
        "name": "fakeUser",
      }
    }
  />
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Cologne
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <React.Fragment>
                <PropertyCard
                  cardMouseEnterHandler={[MockFunction]}
                  offer={
                    Object {
                      "bedrooms": 3,
                      "city": Object {
                        "location": Object {
                          "latitude": 50.938361,
                          "longitude": 6.959974,
                          "zoom": 13,
                        },
                        "name": "Cologne",
                      },
                      "description": "I rent out a very sunny and bright apartment only 7 minutes walking distance to the metro station. The apartment has a spacious living room with a kitchen, one bedroom and a bathroom with mit bath. A terrace can be used in summer.",
                      "goods": Array [
                        "Baby seat",
                        "Fridge",
                        "Towels",
                      ],
                      "host": Object {
                        "avatarUrl": "img/avatar-angelina.jpg",
                        "id": 25,
                        "isPro": true,
                        "name": "Angelina",
                      },
                      "id": 1,
                      "images": Array [
                        "https://es31-server.appspot.com/six-cities/static/hotel/2.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/4.jpg",
                      ],
                      "isFavorite": false,
                      "isPremium": false,
                      "location": Object {
                        "latitude": 50.957361,
                        "longitude": 6.9509739999999995,
                        "zoom": 16,
                      },
                      "maxAdults": 5,
                      "previewImage": "https://es31-server.appspot.com/six-cities/static/hotel/2.jpg",
                      "price": 459,
                      "rating": 3.6,
                      "title": "Penthouse, 4-5 rooms + 5 balconies",
                      "type": "apartment",
                    }
                  }
                />
              </React.Fragment>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Brussels
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <React.Fragment>
                <PropertyCard
                  cardMouseEnterHandler={[MockFunction]}
                  offer={
                    Object {
                      "bedrooms": 5,
                      "city": Object {
                        "location": Object {
                          "latitude": 50.846557,
                          "longitude": 4.351697,
                          "zoom": 13,
                        },
                        "name": "Brussels",
                      },
                      "description": "Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.",
                      "goods": Array [
                        "Laptop friendly workspace",
                      ],
                      "host": Object {
                        "avatarUrl": "img/avatar-angelina.jpg",
                        "id": 25,
                        "isPro": true,
                        "name": "Angelina",
                      },
                      "id": 2,
                      "images": Array [
                        "https://es31-server.appspot.com/six-cities/static/hotel/1.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/20.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/16.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/5.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/10.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/13.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/4.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/8.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/6.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/2.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/15.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/3.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/12.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/18.jpg",
                      ],
                      "isFavorite": false,
                      "isPremium": false,
                      "location": Object {
                        "latitude": 50.869557,
                        "longitude": 4.332697,
                        "zoom": 16,
                      },
                      "maxAdults": 10,
                      "previewImage": "https://es31-server.appspot.com/six-cities/static/hotel/13.jpg",
                      "price": 474,
                      "rating": 2.5,
                      "title": "Waterfront with extraordinary view",
                      "type": "house",
                    }
                  }
                />
              </React.Fragment>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Dusseldorf
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <React.Fragment>
                <PropertyCard
                  cardMouseEnterHandler={[MockFunction]}
                  offer={
                    Object {
                      "bedrooms": 3,
                      "city": Object {
                        "location": Object {
                          "latitude": 51.225402,
                          "longitude": 6.776314,
                          "zoom": 13,
                        },
                        "name": "Dusseldorf",
                      },
                      "description": "Design interior in most sympathetic area! Complitely renovated, well-equipped, cosy studio in idyllic, over 100 years old wooden house. Calm street, fast connection to center and airport.",
                      "goods": Array [
                        "Towels",
                        "Laptop friendly workspace",
                        "Baby seat",
                        "Washer",
                        "Air conditioning",
                        "Breakfast",
                      ],
                      "host": Object {
                        "avatarUrl": "img/avatar-angelina.jpg",
                        "id": 25,
                        "isPro": true,
                        "name": "Angelina",
                      },
                      "id": 3,
                      "images": Array [
                        "https://es31-server.appspot.com/six-cities/static/hotel/10.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/1.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/13.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/4.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/8.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/20.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/12.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/5.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/7.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/9.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/17.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/2.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/11.jpg",
                        "https://es31-server.appspot.com/six-cities/static/hotel/15.jpg",
                      ],
                      "isFavorite": false,
                      "isPremium": false,
                      "location": Object {
                        "latitude": 51.216402,
                        "longitude": 6.758314,
                        "zoom": 16,
                      },
                      "maxAdults": 5,
                      "previewImage": "https://es31-server.appspot.com/six-cities/static/hotel/16.jpg",
                      "price": 474,
                      "rating": 2.4,
                      "title": "Waterfront with extraordinary view",
                      "type": "hotel",
                    }
                  }
                />
              </React.Fragment>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
